### ISSUE TITLE
m4 runtime dependency

### ISSUE THREAD

**Issue by langmm** [conda-forge/bison-feedstock#20]
Issue:

I am currently trying to prepare a package recipe for conda-forge which requires bison for its build. However, I think because m4 is not a runtime dependency, bison is using the system m4 (which is the incorrect version) even when m4 is specified as a build requirement within the recipe or installed within the root environment used to call conda build (I think because the root env has not bearing on the build env and the build m4 is installed after bison). The actual error is

```
/Library/Developer/CommandLineTools/usr/bin/m4: unrecognized option `--gnu'
```

which occurs because the ``--gnu`` flag wasn't introduced until 1.4.11.

```
$ /Library/Developer/CommandLineTools/usr/bin/m4 --version
GNU M4 1.4.6
Copyright (C) 2006 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Written by Rene' Seindal.
```

Why is m4 not a runtime dependency for bison? If there is not a specific reason, I am happy to prepare a pull request to add it. Or is there a work around for forcing bison (installed via conda-forge) to use a different m4 installation?

<br/>
Environment (<code>conda list</code>):
<details>

```
$ conda list
# packages in environment at /Users/langmm/miniconda3:
#
# Name                    Version                   Build  Channel
appnope                   0.1.0                 py37_1000    conda-forge
asn1crypto                0.24.0                py37_1003    conda-forge
backcall                  0.1.0                      py_0    conda-forge
beautifulsoup4            4.8.0                    py37_0    conda-forge
bzip2                     1.0.8                h01d97ff_0    conda-forge
ca-certificates           2019.6.16            hecc5488_0    conda-forge
certifi                   2019.6.16                py37_1    conda-forge
cffi                      1.12.3           py37hccf1714_0    conda-forge
chardet                   3.0.4                 py37_1003    conda-forge
click                     7.0                        py_0    conda-forge
conda                     4.7.11                   py37_0    conda-forge
conda-build               3.18.9                   py37_3    conda-forge
conda-forge-pinning       2019.08.11                    1    conda-forge
conda-package-handling    1.4.1                    py37_0    conda-forge
conda-verify              3.1.1                 py37_1000    conda-forge
cryptography              2.7              py37h212c5bf_0    conda-forge
decorator                 4.4.0                      py_0    conda-forge
entrypoints               0.3                   py37_1000    conda-forge
filelock                  3.0.10                     py_0    conda-forge
flake8                    3.7.8                    py37_1    conda-forge
future                    0.17.1                py37_1000    conda-forge
glob2                     0.7                        py_0    conda-forge
icu                       64.2                 h6de7cb9_0    conda-forge
idna                      2.8                   py37_1000    conda-forge
ipython                   7.7.0            py37h5ca1d4c_0    conda-forge
ipython_genutils          0.2.0                      py_1    conda-forge
jedi                      0.15.1                   py37_0    conda-forge
jinja2                    2.10.1                     py_0    conda-forge
libarchive                3.3.3             h5c473cc_1006    conda-forge
libcxx                    8.0.0                         4    conda-forge
libcxxabi                 8.0.0                         4    conda-forge
libffi                    3.2.1             h6de7cb9_1006    conda-forge
libiconv                  1.15              h01d97ff_1005    conda-forge
liblief                   0.9.0                h2a1bed3_1    conda-forge
libxml2                   2.9.9                h12c6b28_2    conda-forge
lz4-c                     1.8.3             h6de7cb9_1001    conda-forge
lzo                       2.10              h1de35cc_1000    conda-forge
m4                        1.4.18            h1de35cc_1001    conda-forge
markupsafe                1.1.1            py37h1de35cc_0    conda-forge
mccabe                    0.6.1                      py_1    conda-forge
ncurses                   6.1               h0a44026_1002    conda-forge
openssl                   1.1.1c               h01d97ff_0    conda-forge
parso                     0.5.1                      py_0    conda-forge
patch                     2.7.6             h1de35cc_1001    conda-forge
pexpect                   4.7.0                    py37_0    conda-forge
pickleshare               0.7.5                 py37_1000    conda-forge
pip                       19.2.2                   py37_0    conda-forge
pkginfo                   1.5.0.1                    py_0    conda-forge
pkglts                    4.6.1                    pypi_0    pypi
prompt_toolkit            2.0.9                      py_0    conda-forge
psutil                    5.6.3            py37h01d97ff_0    conda-forge
ptyprocess                0.6.0                   py_1001    conda-forge
py-lief                   0.9.0            py37h6d6d4d2_1    conda-forge
pycodestyle               2.5.0                      py_0    conda-forge
pycosat                   0.6.3           py37h1de35cc_1001    conda-forge
pycparser                 2.19                     py37_1    conda-forge
pyflakes                  2.1.1                      py_0    conda-forge
pygments                  2.4.2                      py_0    conda-forge
pyopenssl                 19.0.0                   py37_0    conda-forge
pysocks                   1.7.0                    py37_0    conda-forge
python                    3.7.3                h93065d6_1    conda-forge
python-libarchive-c       2.8                   py37_1004    conda-forge
pytz                      2019.2                     py_0    conda-forge
pyyaml                    5.1.2            py37h01d97ff_0    conda-forge
readline                  8.0                  hcfe32e1_0    conda-forge
requests                  2.22.0                   py37_1    conda-forge
ripgrep                   11.0.2                        0    conda-forge
ruamel_yaml               0.15.71         py37h1de35cc_1000    conda-forge
semver                    2.8.1                    pypi_0    pypi
setuptools                41.0.1                   py37_0    conda-forge
six                       1.12.0                py37_1000    conda-forge
soupsieve                 1.9.2                    py37_0    conda-forge
sqlite                    3.29.0               hb7d70f7_0    conda-forge
tk                        8.6.9             h2573ce8_1002    conda-forge
tqdm                      4.33.0                     py_0    conda-forge
traitlets                 4.3.2                 py37_1000    conda-forge
unidecode                 1.1.1                    pypi_0    pypi
urllib3                   1.25.3                   py37_0    conda-forge
wcwidth                   0.1.7                      py_1    conda-forge
wheel                     0.33.4                   py37_0    conda-forge
xz                        5.2.4             h1de35cc_1001    conda-forge
yaml                      0.1.7             h1de35cc_1001    conda-forge
zlib                      1.2.11            h01d97ff_1005    conda-forge
zstd                      1.4.0                ha9f0a20_0    conda-forge
```
</details>

<br/>
Details about  <code>conda</code> and system ( <code>conda info</code> ):
<details>

```
$ conda info

     active environment : base
    active env location : /Users/langmm/miniconda3
            shell level : 1
       user config file : /Users/langmm/.condarc
 populated config files : /Users/langmm/.condarc
          conda version : 4.7.11
    conda-build version : 3.18.9
         python version : 3.7.3.final.0
       virtual packages : 
       base environment : /Users/langmm/miniconda3  (writable)
           channel URLs : https://conda.anaconda.org/conda-forge/osx-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://repo.anaconda.com/pkgs/main/osx-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/osx-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /Users/langmm/miniconda3/pkgs
                          /Users/langmm/.conda/pkgs
       envs directories : /Users/langmm/miniconda3/envs
                          /Users/langmm/.conda/envs
               platform : osx-64
             user-agent : conda/4.7.11 requests/2.22.0 CPython/3.7.3 Darwin/18.6.0 OSX/10.14.5
                UID:GID : 501:20
             netrc file : None
           offline mode : False

```
</details>

**Comment by mingwandroid** [comment]
AFAIUI (and I could be wrong) `bison` does not call `m4` directly, it calls `flex` which calls `m4`.

The problem is then probably that bison only has a `host` dependency on `flex` when it needs (also?) a `run` dependency on `flex`.

**Comment by mingwandroid** [comment]
Can you commit your WIP so I can look into this?

**Comment by langmm** [comment]
I created a pull request adding flex as a run dependency, but now realize you might have meant commits to the new package recipe. For the new recipe it is tricky, I am simultaneously preparing recipes for 3 interdependent packages from the openalea project and making updates to them to be Python 3 compat. I have provided links to the packages below (in the order they have to be built), but they are very rough and have to be cloned/built independently.

- https://github.com/langmm/deploy
- https://github.com/langmm/sconsx
- https://github.com/langmm/plantgl
